src/charm.py

```diff
@@ -0,0 +1,319 @@
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self.on.amf_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self._default_database.on.database_created, self._on_amf_pebble_ready
+        )
+        self.framework.observe(self._amf_database.on.database_created, self._on_amf_pebble_ready)
+        self.framework.observe(self.on.amf_pebble_ready, self._on_amf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_amf_pebble_ready)
+
+    def _on_amf_pebble_ready(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for AMF container.
+
+        Args:
+            event (PebbleReadyEvent, DatabaseCreatedEvent, NRFAvailableEvent): Juju event
+        """
+        if not self._amf_container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for service to start")
+            event.defer()
+            return
+        for relation in ["fiveg_nrf", "amf-database", "default-database"]:
+            if not self._relation_created(relation):
+                self.unit.status = BlockedStatus(f"Waiting for {relation} relation")
+                return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the default database to be available")
+            event.defer()
+            return
+        if not self._amf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the amf database to be available")
+            event.defer()
+            return
+        if not self._default_database_info:
+            self.unit.status = WaitingStatus("Waiting for default database info to be available")
+            event.defer()
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        content = self._render_config_file(
+            ngapp_port=NGAPP_PORT,
+            sctp_grpc_port=SCTP_GRPC_PORT,
+            sbi_port=SBI_PORT,
+            nrf_url=self._nrf_requires.nrf_url,
+            amf_url=self._amf_hostname,
+            default_database_name=DEFAULT_DATABASE_NAME,
+            amf_database_name=AMF_DATABASE_NAME,
+            database_url=self._default_database_info["uris"].split(",")[0],
+        )
+        if not self._config_file_is_pushed or not self._config_file_content_matches(
+            content=content
+        ):
+            self._push_config_file(
+                content=content,
+            )
+        self._amf_container.add_layer("amf", self._amf_pebble_layer, combine=True)
```

Comment:
Is there any reason why we don't handle the layer and the workload service the same way we did in Charmed Magma? I mean do we want to add the layer every time even if it doesn't change? And don't we need to restart the service?